src/workerd/util/sqlite.c++

```diff
@@ -51,14 +51,18 @@ namespace workerd {
     int _ec = code; \
     /* SQLITE_MISUSE doesn't put error info on the database object, so check it separately */ \
     KJ_ASSERT(_ec != SQLITE_MISUSE, "SQLite misused: " #code, ##__VA_ARGS__); \
-    SQLITE_REQUIRE(_ec == SQLITE_OK, sqlite3_errmsg(db), " " #code, ##__VA_ARGS__); \
+    int offset = sqlite3_error_offset(db); \
+    bool showOffsetMsg = offset != -1; \
+    SQLITE_REQUIRE(_ec == SQLITE_OK, kj::str(sqlite3_errmsg(db), showOffsetMsg ? kj::str(" at offset ", offset) : kj::str()), " " #code, ##__VA_ARGS__); \
```

Comment:
Here and in line 65, you can replace the `kj::str()` with `""_kj`, which should avoid allocating an empty String and instead use a StringPtr pointing to the empty string.

Reply:
Awesome, I was wondering if there was a better way than `kj::str()`. Thanks!

Reply:
FWIW you can also use `nullptr`, which converts to `kj::String` without an allocation. If you use `""_kj`, I think that causes the `?:` overall to evaluate to type `kj::StringPtr` instead of `kj::String`, which is probably fine since the string is used immediately on this line, but would be a problem if assigning to a local variable (you'd end up with a dangling pointer).

On another somewhat-related note, it looks like this whole `kj::str()` expression will be evaluated twice by the `SQLITE_REQUIRE` macro, duplicating all the allocations. So assigning to a local variable before calling `SQLITE_REQUIRE` might save cycles.

Then again, this is an error path, so performance is not relevant.

Reply:
In the scenario of using `nullptr` and a local variable, does this look right?

```cpp
bool showOffsetMsg = offset != -1; \
kj::String msg = kj::str(sqlite3_errmsg(db), showOffsetMsg ? kj::str(" at offset ", offset) : nullptr); \
```

Reply:
Yes. Though actually at this point it might be more readable to do this:

```
kj::String msg = kj::str(sqlite3_errmsg(db));
if (offset != -1) {
  msg = kj::str(msg, " at offset ", offset);
}
```

And it's the same number of allocations anyway...